### ISSUE TITLE
Implement `__eq__` for IRFs

### ISSUE THREAD

**Issue by maxnoe** [gammapy/gammapy#3534]
**Is your feature request related to a problem? Please describe.**

Currently `IRF` does not implement `__eq__`, so it just tests object identity. Would make e.g. writing unit tests easier.

**Describe the solution you'd like**
Implement `__eq__` for IRF.

**Comment by AtreyeeS** [comment]
Hi @maxnoe What would you like to test the equivalency for? `data`, `unit`, and the `axes` like in maps, or also some extra info like meta etc?

**Comment by adonath** [comment]
I agree this is a useful feature and will lead to nice and readable code. In practice one might need to allow for a certain tolerance, which is the downside of the `__eq__`, where one cannot pass additional parameters. On the other hand one can certainly work with a good default choice...

**Comment by bkhelifi** [comment]
We have 4 types of IRFs, so we might need (at least) 4 __eq__ for each type.
For `EffectiveAreaTable2D`, we can compare the 2 axis and the data
For `BackgroundIRF`, the axis and the data.
For the PSF classes, one might need a different comparison for each PSF class.
For `EnergyDispersion2D`, one can compare the 3 axis and the data...
Do I analyse correctly the 4 tasks, one per IRF type?

**Comment by adonath** [comment]
@bkhelifi In practice I don't think we need a different implementation for the different IRF types, except maybe for the parametric PSFs. One pattern that could be useful is to add either a `IRF.compare(other, atol, rtol)` or maybe `IRF.allclose(r, atol, rtol)`. This would allow for configurable comparison and maybe provide a more meaningful printout of what is different. The `__eq__` method then just calls into `.compare()` with a fixed pre-defined (class level defined) precision.

**Comment by adonath** [comment]
And if we implement it for `IRF`, we should do it for `Map` as well.

**Comment by maxnoe** [comment]
@adonath with `Map` I would be probably expect something else. A map is something like an ndarray with named axes, right?

So I would expect it to behave like it, i.e. do such comparisons element wise and return a boolean map? Raise an error if map axes differ?

**Comment by registerrier** [comment]
> @adonath with `Map` I would be probably expect something else. A map is something like an ndarray with named axes, right?
> 
> So I would expect it to behave like it, i.e. do such comparisons element wise and return a boolean map of the same dimensions? Raise an error if map axes differ?

@adonath it actually already exists on `Map` and behaves exactly as @maxnoe suggests (raise an error if map geometries differ).

https://github.com/gammapy/gammapy/blob/e8b7012df864edbcf355bf32d0468264f99cdb8a/gammapy/maps/core.py#L1586

**Comment by adonath** [comment]
Thanks @maxnoe and @registerrier! Indeed I forgot that we implemented this already and the way it's implemented is the correct one. A method like `Map.allclose(other, atol, rtol)` might still make sense.

However this raises to me the question, whether `IRF.__eq__` should maybe behave exactly the same way. I.e. return the element wise comparison of an IRF with the same axes. However I see that an IRF containing booleans deviates at lot from the original concept....

But an IRF is also mostly a ndarray with named axes, such that some operations e.g. computing the weighted average of two different IRFs of the same kind, like:
```python3
aeff_mean = (aeff_1 * w_1 + aeff_2 * w_2) / (w_1 + w_2)
```
...are well defined and and make sense. And even an operation like:
```python3
aeff = aeff_1 * (aeff_1 > some_value)
```
...is reasonably meaningful. But this definitely extends the original concept of an IRF and requires a lot more thought on which operations and arithmetics to support. In this sense a solution like `IRF.allclose(other, )` or `IRF.compare()` is probably more flexible in defining a reasonable behavior.

**Comment by maxnoe** [comment]
@adonath only for aeff and background. These are the only IRFs with absolute values, everything else is probability distributions and this simple interpolation / averaging will produce nonsensical results.

This is a big topic currently in pyirf / LST / CTA, see https://github.com/cta-observatory/pyirf/issues/161

**Comment by registerrier** [comment]
Closed after #3935 and #3887
